// node-iframe/dist/iframe.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.fetchFrame = exports.configureTemplates = exports.configureResourceControl = exports.configureAgent = exports.appSourceConfig = void 0;
var isomorphic_unfetch_1 = __importDefault(require("isomorphic-unfetch"));
var cheerio_1 = require("cheerio");
var utils_1 = require("./utils");
var templates_1 = require("./templates");
exports.configureTemplates = templates_1.configureTemplates;
var config_1 = require("./config");
var appSourceConfig = config_1.defaultConfig;
exports.appSourceConfig = appSourceConfig;
var httpAgent;
var httpsAgent;
var agentConfigured = false;
var agent = agentConfigured
    ? function (_parsedURL) {
        if (_parsedURL.protocol == "http:" && setAgent(true)) {
            return httpAgent;
        }
        else if (setAgent(false)) {
            setAgent(false);
            return httpsAgent;
        }
    }
    : undefined;
function setAgent(http) {
    try {
        if (http && !httpAgent) {
            var transport = require("http");
            httpAgent = new transport.Agent({
                rejectUnauthorized: false
            });
        }
        else if (!httpsAgent) {
            var transport = require("https");
            httpsAgent = new transport.Agent({
                rejectUnauthorized: false
            });
        }
        return true;
    }
    catch (e) {
        console.error(e);
        return false;
    }
}
function configureAgent() {
    if (!agentConfigured) {
        try {
            if (typeof window === "undefined" && !agentConfigured) {
                agentConfigured = true;
            }
        }
        catch (e) {
            console.error(e);
        }
    }
}
exports.configureAgent = configureAgent;
var mutateSource = function (_a, url, $html, headers) {
    var _b = _a.src, src = _b === void 0 ? "" : _b, key = _a.key;
    return __awaiter(void 0, void 0, void 0, function () {
        var res, source, e_1;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!(src && src[0] === "/")) return [3, 6];
                    _c.label = 1;
                case 1:
                    _c.trys.push([1, 5, , 6]);
                    return [4, isomorphic_unfetch_1["default"](url + "/" + src, {
                            headers: headers,
                            agent: agent
                        })];
                case 2:
                    res = _c.sent();
                    return [4, res.text()];
                case 3:
                    source = _c.sent();
                    return [4, $html(key).html(source)];
                case 4:
                    _c.sent();
                    return [3, 6];
                case 5:
                    e_1 = _c.sent();
                    console.error(e_1);
                    return [3, 6];
                case 6: return [2];
            }
        });
    });
};
function renderErrorHtml(_a) {
    var url = _a.url, server = _a.server, _b = _a.noPage, noPage = _b === void 0 ? false : _b;
    return Object.assign(cheerio_1.load(!url
        ? templates_1.templateModel[templates_1.TemplateType.error]
        : templates_1.templateModel[templates_1.TemplateType.notFound]), server ? { status: Number("" + 40 + (!url || noPage ? 4 : 0)) } : {});
}
function renderHtml(_a, server) {
    var e_2, _b, e_3, _c, e_4, _d;
    var url = _a.url, baseHref = _a.baseHref, config = _a.config, _e = _a.head, head = _e === void 0 ? {} : _e;
    if (server === void 0) { server = false; }
    return __awaiter(this, void 0, void 0, function () {
        var _f, inline, cors, headers, response, html, $html, inlineMutations_2, _g, _h, key, attribute_1, e_2_1, inlineMutations_1, inlineMutations_1_1, com, key, attribute, src, element, e_3_1, _j, _k, key, e_4_1, e_5;
        return __generator(this, function (_l) {
            switch (_l.label) {
                case 0:
                    if (!utils_1.isUrl(url)) {
                        return [2, renderErrorHtml({ url: url, server: server })];
                    }
                    _f = {
                        inline: __assign(__assign({}, appSourceConfig.inline), config === null || config === void 0 ? void 0 : config.inline),
                        cors: __assign(__assign({}, appSourceConfig.cors), config === null || config === void 0 ? void 0 : config.cors)
                    }, inline = _f.inline, cors = _f.cors;
                    headers = __assign(__assign({}, config_1.appHeaders), head);
                    _l.label = 1;
                case 1:
                    _l.trys.push([1, 46, , 47]);
                    return [4, isomorphic_unfetch_1["default"](url, {
                            headers: head,
                            agent: agent
                        })];
                case 2:
                    response = _l.sent();
                    return [4, response.text()];
                case 3:
                    html = _l.sent();
                    $html = cheerio_1.load(html);
                    if (!!!baseHref) return [3, 5];
                    return [4, $html("head").prepend("<base target=\"_self\" href=\"" + url + "\">")];
                case 4:
                    _l.sent();
                    _l.label = 5;
                case 5:
                    inlineMutations_2 = [];
                    _l.label = 6;
                case 6:
                    _l.trys.push([6, 12, 13, 18]);
                    _g = __asyncValues(Object.keys(inline));
                    _l.label = 7;
                case 7: return [4, _g.next()];
                case 8:
                    if (!(_h = _l.sent(), !_h.done)) return [3, 11];
                    key = _h.value;
                    if (!inline[key]) return [3, 10];
                    attribute_1 = "src";
                    return [4, $html(key).attr(attribute_1, function (_, src) {
                            if (src) {
                                inlineMutations_2.push({ key: key, attribute: attribute_1, src: src });
                            }
                            return src;
                        })];
                case 9:
                    _l.sent();
                    _l.label = 10;
                case 10: return [3, 7];
                case 11: return [3, 18];
                case 12:
                    e_2_1 = _l.sent();
                    e_2 = { error: e_2_1 };
                    return [3, 18];
                case 13:
                    _l.trys.push([13, , 16, 17]);
                    if (!(_h && !_h.done && (_b = _g["return"]))) return [3, 15];
                    return [4, _b.call(_g)];
                case 14:
                    _l.sent();
                    _l.label = 15;
                case 15: return [3, 17];
                case 16:
                    if (e_2) throw e_2.error;
                    return [7];
                case 17: return [7];
                case 18:
                    _l.trys.push([18, 25, 26, 31]);
                    inlineMutations_1 = __asyncValues(inlineMutations_2);
                    _l.label = 19;
                case 19: return [4, inlineMutations_1.next()];
                case 20:
                    if (!(inlineMutations_1_1 = _l.sent(), !inlineMutations_1_1.done)) return [3, 24];
                    com = inlineMutations_1_1.value;
                    key = com.key, attribute = com.attribute, src = com.src;
                    element = key + "[" + attribute + "=\"" + src + "\"]";
                    return [4, mutateSource({ key: element, src: src }, url, $html, headers)];
                case 21:
                    _l.sent();
                    return [4, $html(element).removeAttr(attribute)];
                case 22:
                    _l.sent();
                    _l.label = 23;
                case 23: return [3, 19];
                case 24: return [3, 31];
                case 25:
                    e_3_1 = _l.sent();
                    e_3 = { error: e_3_1 };
                    return [3, 31];
                case 26:
                    _l.trys.push([26, , 29, 30]);
                    if (!(inlineMutations_1_1 && !inlineMutations_1_1.done && (_c = inlineMutations_1["return"]))) return [3, 28];
                    return [4, _c.call(inlineMutations_1)];
                case 27:
                    _l.sent();
                    _l.label = 28;
                case 28: return [3, 30];
                case 29:
                    if (e_3) throw e_3.error;
                    return [7];
                case 30: return [7];
                case 31: return [4, $html("[src=\"undefined\"]").removeAttr("src")];
                case 32:
                    _l.sent();
                    _l.label = 33;
                case 33:
                    _l.trys.push([33, 39, 40, 45]);
                    _j = __asyncValues(Object.keys(cors));
                    _l.label = 34;
                case 34: return [4, _j.next()];
                case 35:
                    if (!(_k = _l.sent(), !_k.done)) return [3, 38];
                    key = _k.value;
                    if (!cors[key]) return [3, 37];
                    return [4, $html(key).attr("crossorigin", cors[key])];
                case 36:
                    _l.sent();
                    _l.label = 37;
                case 37: return [3, 34];
                case 38: return [3, 45];
                case 39:
                    e_4_1 = _l.sent();
                    e_4 = { error: e_4_1 };
                    return [3, 45];
                case 40:
                    _l.trys.push([40, , 43, 44]);
                    if (!(_k && !_k.done && (_d = _j["return"]))) return [3, 42];
                    return [4, _d.call(_j)];
                case 41:
                    _l.sent();
                    _l.label = 42;
                case 42: return [3, 44];
                case 43:
                    if (e_4) throw e_4.error;
                    return [7];
                case 44: return [7];
                case 45:
                    if (server) {
                        $html.status = 200;
                    }
                    return [2, $html];
                case 46:
                    e_5 = _l.sent();
                    console.error(e_5);
                    return [3, 47];
                case 47: return [2, renderErrorHtml({ url: url, server: server, noPage: true })];
            }
        });
    });
}
function fetchFrame(model) {
    return __awaiter(this, void 0, void 0, function () {
        var $html;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4, renderHtml(model, typeof process !== "undefined")];
                case 1:
                    $html = _a.sent();
                    return [2, $html === null || $html === void 0 ? void 0 : $html.html()];
            }
        });
    });
}
exports.fetchFrame = fetchFrame;
function configureResourceControl(appConfig) {
    exports.appSourceConfig = appSourceConfig = Object.assign({}, config_1.defaultConfig, {
        cors: __assign(__assign({}, config_1.defaultCorsConfig), appConfig === null || appConfig === void 0 ? void 0 : appConfig.cors),
        inline: __assign(__assign({}, config_1.defaultInlineConfig), appConfig === null || appConfig === void 0 ? void 0 : appConfig.inline)
    });
}
exports.configureResourceControl = configureResourceControl;
function createIframe(_req, res, next) {
    var _this = this;
    res.createIframe = function (model) { return __awaiter(_this, void 0, void 0, function () {
        var $html;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4, renderHtml(model, true)];
                case 1:
                    $html = _a.sent();
                    res.status(($html === null || $html === void 0 ? void 0 : $html.status) || 200).send($html.html());
                    return [2];
            }
        });
    }); };
    next();
}
exports["default"] = createIframe;
